woocommerce/woocommerce
# Getting the Product Collection location/context

Getting the Product Collection location/context

This PR includes creating a `useGetLocation` hook that recognizes the location and context of Product Collection block (and potentially other blocks).

Examples:

_Note: the below screenshot were taken for demo purposes. The block doesn't render that data in the Editor in this PR!_

In Specific category template (Category: Clothing) | In Single Product template | In post/page
-- | -- | --
<img width="1184" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/e89435a2-d374-41eb-b3c6-7fd5e5ec8cab"> | <img width="1184" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/5adc6ee4-d297-4b2b-8e5b-dc6c791c1f22"> | <img width="1181" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/fe454674-6124-471b-8940-758fbc8cbb57">

### Prerequisites to testing:
1. To verify the location, go to Developer Tools -> Network tab
2. Search for: `product isproductcollection`

### Testing steps:
1. Try adding Product Collection block (choose some collection, for example, "Featured") in multiple places (example scenarios are provided in the next step) and verify the proper payload is sent with the request by checking Payload tab.
- Example from Single Product template:
  - `location[type]` is `product`
  - productId is unknown, hence `location[sourceData][productId]` is empty:
 
![Screen Shot 2024-02-09 at 17 43 51 PM](https://github.com/woocommerce/woocommerce/assets/20098064/813fa263-e9b6-42f9-999e-911f061c625c)

2. Here are some example scenarios, but try to break it as well by trying something else:

Scenario | Output
-- | --
Products by Category/Tag/Attribute template | `location[type]: 'archive'` <br/> `location[sourceData][taxonomy]: 'product_cat'` <br/>`location[sourceData][taxonomy]:`
Category: SPECIFIC_CATEGORY template | `location[type]: 'archive'` <br/> `location[sourceData][taxonomy]: 'product_cat'` <br/>`location[sourceData][taxonomy]: SPECIFIC_CATEGORY_ID`
Product Collection in Single Product block in Category: SPECIFIC_CATEGORY template | `location[type]: 'product'`<br/> `location[sourceData][productId]: SPECIFIC_PRODUCT_ID`
Product Collection in Single Product: SPECIFIC_PRODUCT template | `location[type]: 'product'` <br/> `location[sourceData][productID]: PRODUCT_ID`
In the above scenario, add Single Product block and Product Collection inside | `location[type]: 'product'`<br/> `location[sourceData][productID]: PRODUCT_ID_FROM_SINGLE_PRODUCT_BLOCK }`
Product Collection in Mini Cart Contents | `location[type]: 'cart'`
Page/Post | `location[type]: 'generic'`

5. Location type matches the closest context and ID is provided if available.

### Additional instructions - how to add custom templates or modify some blocks

#### Create "Single Product: SPECIFIC_PRODUCT template"
1. Go to Editor -> Templates -> "+" button -> Single Item: Product
2. Choose some product

#### Adding Product Collection inside Single Product block
1. Add "Single Product" block from the inserter
<img width="301" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/da1b585e-5fff-4cf7-b03b-2998b130e482">

2. Choose some product and click "Done"
3. Open List View of blocks
4. Expand Single Product block -> Columns -> Column
5. Click "Three dots" menu next to some of the inner blocks and click "Add After"
6. Insert Product Collection and choose some collection, for example "Featured"

<img width="576" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/7f0f939d-7b7f-4067-b708-1e3c9b3c034b">

#### Product Collection in Mini Cart Contents

1. Go to Editor -> Patterns -> Scroll to the very bottom and choose "Mini Cart" -> "Mini Cart"

<img width="718" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/92e152fd-650e-4d0e-8bd3-c5cc74736fab">

2. Edit template part
3. Open List view of blocks
4. Expand Mini Cart Contents -> Filled Mini Cart -> Mini Cart Items
5. Click "Three dots" menu next to some of the inner blocks and click "Add After"
6. Insert Product Collection and choose some collection, for example "Featured"




Closes: https://github.com/woocommerce/woocommerce/issues/43694
